File: 2024/07.py
import time


def parse(src):
    return [(int(l.split(': ')[0]), tuple(map(int, l.split(': ')[1].split(' ')))) for l in src.splitlines()]


def part1(src):
    def helper(total, acc, inputs):
        if not inputs:
            return total == acc
        v, *rest = inputs
        return helper(total, acc + v, rest) or helper(total, acc * v, rest)

    x = parse(src)
    return sum(total for total, inputs in parse(src) if helper(total, inputs[0], inputs[1:]))


def part2(src):
    def helper(total, acc, inputs):
        if not inputs:
            return total == acc
        v, *rest = inputs
        return (helper(total, acc + v, rest) or
                helper(total, acc * v, rest) or
                helper(total, int(str(acc) + str(v)), rest))

    return sum(total for total, inputs in parse(src) if helper(total, inputs[0], inputs[1:]))


SAMPLE = """\
190: 10 19
3267: 81 40 27
83: 17 5
156: 15 6
7290: 6 8 6 15
161011: 16 10 13
192: 17 8 14
21037: 9 7 18 13
292: 11 6 16 20
"""

assert part1(SAMPLE) == 3749
assert part2(SAMPLE) == 11387

with open('in/07.txt', 'rt') as f:
    data = f.read()

print(part1(data))
start = time.time()
print(part2(data))
print(time.time() - start)

File: c/2024/07.c
#include <stdio.h>
#include <stdlib.h>

#include "../util.h"


const char *SAMPLE =
"190: 10 19\n"
"3267: 81 40 27\n"
"83: 17 5\n"
"156: 15 6\n"
"7290: 6 8 6 15\n"
"161011: 16 10 13\n"
"192: 17 8 14\n"
"21037: 9 7 18 13\n"
"292: 11 6 16 20\n";

static long common(const char *src, int (*helper)(long, long, char *, char **)) {
    long sum = 0;
    while (*src) {
        char *p;
        long int total = strtol(src, &p, 10);
        p++;
        long int first = strtol(p, &p, 10);
        if (helper(total, first, p, &p)) {
            sum += total;
        }
        src = p;
    }
    return sum;
}

static int helper1(long total, long acc, char *src, char **next) {
    if (*src == '\n') {
        *next = src + 1;
        return total == acc;
    }
    char *p;
    long v = strtol(src, &p, 10);
    return helper1(total, acc + v, p, next) || helper1(total, acc * v, p, next);
}

static long long part1(const char *src) {
    return common(src, helper1);
}

static int helper2(long total, long acc, char *src, char **next) {
    if (*src == '\n') {
        *next = src + 1;
        return total == acc;
    }
    char *p;
    long v = strtol(src, &p, 10);
    if (helper2(total, acc + v, p, next) || helper2(total, acc * v, p, next)) {
        return 1;
    }
    char buf[40];
    sprintf(buf, "%ld%ld", acc, v);
    return helper2(total, strtol(buf, NULL, 10), p, next);
}

static long part2(const char *src) {
    return common(src, helper2);
}

int main(int argc, char *argv[]) {
    assert(part1(SAMPLE) == 3749);
    assert(part2(SAMPLE) == 11387);
    char *src = read_input("2024/in/07.txt");
    printf("%ld\n", part1(src));
    printf("%ld\n", part2(src));
    free(src);
}
